cirq/protocols/measurement_key_protocol.py

```diff
@@ -67,28 +84,62 @@ def measurement_key(val: Any, default: Any = RaiseTypeErrorIfNotProvided):
     Raises:
         TypeError: `val` doesn't have a _measurement_key_ method (or that method
             returned NotImplemented) and also no default value was specified.
+        ValueError: `val` has multiple measurement keys.
     """
-    getter = getattr(val, '_measurement_key_', None)
-    result = NotImplemented if getter is None else getter()
-    if result is not NotImplemented:
-        return result
+    result = measurement_keys(val)
+
+    if len(result) == 1:
+        return result[0]
+
+    if len(result) > 1:
+        raise ValueError(f'Got multiple measurement keys ({result!r}) '
+                         f'from {val!r}.')
 
     if default is not RaiseTypeErrorIfNotProvided:
         return default
 
-    if getter is None:
-        raise TypeError(
-            "object of type '{}' has no _measurement_key_ method.".format(
-                type(val)))
+    raise TypeError(f"object of type '{type(val)}' had no measurement keys.")
+
+
+def measurement_keys(val: Any, *,
+                     allow_decompose: bool = True) -> Tuple[str, ...]:
+    """Gets the measurement keys of measurements within the given value.
+
+    Args:
+        val: The value which has the measurement key..
+        allow_decompose: Defaults to True. When true, composite operations that
+            don't directly specify their measurement keys will be decomposed in
+            order to find measurement keys within the decomposed operations. If
+            not set, composite operations will appear to have no measurement
+            keys.
+
+    Returns:
+        The measurement keys of the value. If the value has no measurement,
+        the result is the empty tuple.
+    """
+    getter = getattr(val, '_measurement_keys_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is not NotImplemented and result is not None:
```

Comment:
nit:

```python
if result not in (NotImplemented, None):
    ...
```

Reply:
Acknowledged. That constructs a tuple as part of the check. More expensive.